File: solutions/python/robot-simulator/1/robot_simulator.py
# Globals for the directions
EAST = 'east'
NORTH = 'north'
WEST = 'west'
SOUTH = 'south'


class Robot:
    def __init__(self, direction=NORTH, x_pos=0, y_pos=0):
        self.direction = direction
        self.coordinates = (x_pos, y_pos)

    def move(self, steps):
        for step in steps:
            if step == 'R':
                self.turn_right()
            elif step == 'L':
                self.turn_left()
            elif step == 'A':
                self.advance()

    def turn_right(self):
        self.direction = {
            NORTH: EAST,
            EAST: SOUTH,
            SOUTH: WEST,
            WEST: NORTH
        }[self.direction]

    def turn_left(self):
        self.direction = {
            NORTH: WEST,
            WEST: SOUTH,
            SOUTH: EAST,
            EAST: NORTH
        }[self.direction]

    def advance(self):
        x, y = self.coordinates
        if self.direction == NORTH:
            self.coordinates = (x, y + 1)
        elif self.direction == SOUTH:
            self.coordinates = (x, y - 1)
        elif self.direction == EAST:
            self.coordinates = (x + 1, y)
        elif self.direction == WEST:
            self.coordinates = (x - 1, y)

File: solutions/typescript/robot-simulator/1/robot-simulator.ts
export class InvalidInputError extends Error {
  constructor(message: string) {
    super()
    this.message = message || 'Invalid Input'
  }
}

type Direction = 'north' | 'east' | 'south' | 'west'
type Coordinates = [number, number]

export class Robot {
  #bearing: Direction = 'north'
  #coordinates: Coordinates = [0, 0]

  get bearing(): Direction {
    return this.#bearing
  }

  get coordinates(): Coordinates {
    return this.#coordinates
  }

  place({direction, x, y}: { x: number; y: number; direction: string }) {
    if (!['north', 'east', 'south', 'west'].includes(direction)) {
      throw new InvalidInputError('Invalid Input')
    }
    this.#bearing = direction as Direction
    this.#coordinates = [x, y]
  }

  evaluate(instructions: string) {
    for (const instruction of instructions) {
      switch (instruction) {
        case 'L':
          this.turnLeft()
          break
        case 'R':
          this.turnRight()
          break
        case 'A':
          this.advance()
          break
      }
    }
  }

  turnRight: () => void = () => {
    const dirs = ['north', 'east', 'south', 'west', 'north']
    this.#bearing = dirs[dirs.indexOf(this.#bearing) + 1] as Direction
  }

  turnLeft: () => void = () => {
    const dirs = ['north', 'east', 'south', 'west', 'north']
    this.#bearing = dirs[dirs.lastIndexOf(this.#bearing) - 1] as Direction
  }

  advance: () => void = () => {
    const v = ['north', 'south']
    const h = ['east', 'west']
    const [x, y] = this.#coordinates
    if (v.includes(this.#bearing)) {
      this.#coordinates[1] = y + (this.#bearing === 'north' ? 1 : -1)
    } else {
      this.#coordinates[0] = x + (this.#bearing === 'east' ? 1 : -1)
    }
  }
}
